Updates for the security section (#168)
neo4j/docs-graphql: modules/ROOT/pages/security/authentication.adoc

= Authentication
:description: This page describes how to set up authentication features in the Neo4j GraphQL Library.

The GraphQL Library offers the `@authentication` directive to configure authentication for certain operations and for different parts of your schema.

> [NOTE]

Reviewer: ```suggestion
[IMPORTANT]
====
```

> Explicit authentication, configured with the `@authentication` directive, is only ever evaluated during Cypher translation time, and unauthenticated requests with queries requiring authentication never reach the database.

Reviewer: ```suggestion
Explicit authentication, configured with the `@authentication` directive, is only ever evaluated during Cypher translation time.
Unauthenticated requests with queries requiring authentication never reach the database.
====
```

== Operations

Authentication can be configured to only be validated on certain operations:

* `CREATE`
* `READ`
* `AGGREGATE`
* `UPDATE`
* `DELETE`
* `CREATE_RELATIONSHIP`
* `DELETE_RELATIONSHIP`
* `SUBSCRIBE`

> 

Reviewer: ```suggestion
```

For instance, to only require authentication for the update or deletion of a user:

[source, graphql, indent=0]
----
type User @authentication(operations: [UPDATE, DELETE]) {
    id: ID!
    name: String!
    password: String!
}
----

> [NOTE]

Reviewer: ```suggestion
[NOTE]
====
```

> If there is no `operations` argument with a list of operations, the GraphQL Library treats the authentication configuration as if the full list of operations had been provided.

Reviewer: ```suggestion
In case there is no `operations` argument with a list of operations, the GraphQL Library treats the authentication configuration as if the full list of operations had been provided.
====
```

> 

Reviewer: ```suggestion
```

== Scope

> 

Reviewer: ```suggestion
```

=== Global authentication

> Athentication can be applied to the entire schema.

Reviewer: ```suggestion
Authentication can be applied to the entire schema.
```

This ensures authentication is checked for every matching request.

> Extend the schema:

Reviewer: I think this sequence of sentences is a bit confusing. Do you mean that: "The `operations` and `jwt` arguments can also be used in case the directive is applied to a schema extension, which can be achieved like this:"?

Author: hm... this is what line 53 says, it's different from what lines 43 -> 46 want to say (+ probably a missing example for line 53)

Reviewer: Ok, so never mind, maybe I didn't understand it correctly 😁

[source, graphql, indent=0]
----
extend schema @authentication
----

The `operations` and `jwt` arguments can also be used when the directive is applied to a schema extension.

> // ^ should we add links or provide examples? 

Reviewer: I would say yes

[IMPORTANT]
====
Explicit authentication, configured with the `@authentication` directive, is only ever evaluated during Cypher translation time.
Unauthenticated requests with queries requiring authentication never reach the database.
====

== Operations

Authentication can be configured to only be validated on certain operations:

* `CREATE`
* `READ`
* `AGGREGATE`
* `UPDATE`
* `DELETE`
* `CREATE_RELATIONSHIP`
* `DELETE_RELATIONSHIP`
* `SUBSCRIBE`

For instance, to only require authentication for the update or deletion of a user:

[source, graphql, indent=0]
----
type User @authentication(operations: [UPDATE, DELETE]) {
    id: ID!
    name: String!
    password: String!
}
----

[NOTE]
====
In case there is no `operations` argument with a list of operations, the GraphQL Library treats the authentication configuration as if the full list of operations had been provided.
====

== Scope

=== Global authentication

Authentication can be applied to the entire schema.
This ensures authentication is checked for every matching request.

Extend the schema:

[source, graphql, indent=0]
----
extend schema @authentication
----

> The `operations` and `jwt` arguments can also be used when the directive is applied to a schema extension.
> 
> // ^ should we add links or provide examples? 

Author: ```suggestion
The `operations` and `jwt` arguments can also be used when the directive is applied to a schema extension, for example:

[source, graphql, indent=0]
----
extend schema @authentication(operations: [UPDATE, DELETE], jwt: { roles_INCLUDES: "admin" })
----
```
